eclass/kernel-install.eclass

```diff
@@ -660,13 +661,65 @@ kernel-install_extract_from_uki() {
 	local extract_type=${1}
 	local uki=${2}
 	local out=${3}
+	local out_temp=${T}/${extract_type}-section-dumped
 
 	# objcopy overwrites input if there is no output, dump the output in T.
 	# We unfortunately cannot use /dev/null here
 	$(tc-getOBJCOPY) "${uki}" "${T}/dump.efi" \
-		--dump-section ".${extract_type}=${out}" ||
-		die "Failed to extract ${extract_type}"
-	chmod 644 "${out}" || die
+		--dump-section ".${extract_type}=${out_temp}" ||
+			die "Failed to extract ${extract_type}"
+
+	# Sanity checks for kernel images
+	if [[ ${extract_type} == linux ]] &&
+		{ ! in_iuse secureboot || use secureboot ;}
+	then
+		# Extract the used SECUREBOOT_SIGN_CERT to verify the kernel image
+		local cert=${T}/pcrpkey
+		kernel-install_extract_from_uki pcrpkey "${uki}" "${cert}"
+		if [[ $(head -n1 "${cert}") != "-----BEGIN CERTIFICATE-----" ]]; then
+			# This is a DER format certificate, convert it to PEM
+			openssl x509 \
+				-inform DER -in "${cert}" \
+				-outform PEM -out "${cert}" ||
+					die "Failed to convert pcrpkey to PEM format"
+		fi
+
+		# Check if the signature on the UKI is valid
+		sbverify --cert "${cert}" "${uki}" ||
+			die "ERROR: UKI signature is invalid"
+
+		# Check if the signature on the kernel image is valid
+		local sbverify_err=$(
+			sbverify --cert "${cert}" "${out_temp}" 2>&1 >/dev/null
```

Comment:
```suggestion
			sbverify --cert "${cert}" "${out_temp}" &>/dev/null
```

We're using bash.

Reply:
This doesn't work, sbverify puts this on stderr:
```
warning: data remaining[20364272 vs 20364288]: gaps between PE/COFF sections?
Hash doesn't match image
got:       c0e71e3fcd337008d7f78d19b5736b73b8fa51d5230911677e953549c8b94532
expecting: c58aecd021d2720206f7baa81a7a4b66317c5fcb91e7a72a794ca9e02336f233
Image fails hash check
```
And this on stdout:
```
Signature verification failed
```

We want to capture stderr and store it, we don't care about stdout so `2>&1 >/dev/null` redirects stderr to stdout and stdout to null. Which gives us an empty variable on success and stores the error on failure.

Reply:
Note that `2>&1 >/dev/null` is not the same as `>/dev/null 2>&1`, the later is equivalent to `&>/dev/null`, the former is not.

Reply:
Ah, sorry, looked at it wrong. You're right, of course.